### PR TITLE
[core][rdt][cgraphs] Only allocate tensor without initializing/writing on recv

### PR DESCRIPTION
**Pull request by dayshah**
## Why are these changes needed?
`torch.zeros` allocates memory and then writes zeros to it. `torch.empty` only allocates the memory, so you'll get whatever existed in memory there before but don't pay the cost of writing which is ok because our recv will do the actual writing here. Did some benchmarks on cpu memory and gpu memory on an aws g4dn.12xlarge. Did them 10-20 times, ~2000x on cpu, ~10x on gpu. The gpu benchmarks did need the warmup loop before for empty to actually get the 10x gains.

Making this change for rdt recv's, cgraphs recv's, and the nccl benchmarking recv in our cgraph benchmarks.

### CPU Memory Benchmark
```
start_time = time.perf_counter()
tensor = torch.empty(10000 * 1024 * 1024, dtype=torch.uint8)
# tensor = torch.zeros(10000 * 1024 * 1024, dtype=torch.uint8)
print(time.perf_counter() - start_time)

# Zeros: 0.44489038293249905
# Empty: 0.00016737193800508976
```

### GPU Memory Benchmark
```
# warm up context/allocator
for _ in range(10):
    _ = torch.empty(1, device="cuda")

start_time = time.perf_counter()
tensor = torch.empty(10000 * 1024 * 1024, dtype=torch.uint8, device="cuda")
# tensor = torch.zeros(10000 * 1024 * 1024, dtype=torch.uint8, device="cuda")
print(time.perf_counter() - start_time)

# Zeros: 0.008416594937443733
# Empty: 0.0006318900268524885
```

### CODE REVIEW

**Review by edoakes — COMMENTED** [review]
should rename `_torch_zeros_allocator`

**Comment by dayshah** [comment]
> should rename `_torch_zeros_allocator`

updated

**Comment by edoakes** [comment]
`recv` is always guaranteed to write the full tensor?

**Comment by dayshah** [comment]
> `recv` is always guaranteed to write the full tensor?

it'll usually hang if we don't, in the future we'll abort in which case we'll throw the output tensor away

it's recommended to use empty, torch distributed docs use empty for output tensors too - http://docs.pytorch.org/docs/stable/distributed.html#torch.distributed.all_to_all_single

stanford cs 336 does the same - https://stanford-cs336.github.io/spring2025-lectures/?trace=var%2Ftraces%2Flecture_08.json&source=lecture_08.py&line=378

unrelated dashboard test failure

**Review by edoakes — APPROVED** [review]
stanfurd
